soroban-env-host/src/e2e_invoke.rs

```diff
@@ -129,14 +134,57 @@ pub struct LedgerEntryLiveUntilChange {
     pub new_live_until_ledger: u32,
 }
 
+// Builds a set for metered lookups of keys for entries that were restored from
+// the archived state.
+// This returns an `Option` instead of an empty set because most of the
+// invocations won't have this populated and it makes sense to not run
+// unnecessary metered work for them (even empty map lookups have some cost
+// charged).
+fn build_restored_key_set(
+    budget: &Budget,
+    resources: &SorobanResources,
+    resource_ext: &SorobanTransactionDataExt,
+) -> Result<Option<RestoredKeySet>, HostError> {
+    match resource_ext {
+        SorobanTransactionDataExt::V0 => Ok(None),
+        SorobanTransactionDataExt::V1(soroban_resources_ext_v0) => {
+            if soroban_resources_ext_v0.archived_soroban_entries.is_empty() {
+                return Ok(None);
+            }
+            let rw_footprint = &resources.footprint.read_write;
+            let mut key_set = RestoredKeySet::default();
+            for e in soroban_resources_ext_v0.archived_soroban_entries.iter() {
+                key_set = key_set.insert(
+                    Rc::new(
+                        rw_footprint
+                            .get(*e as usize)
```

Comment:
Do we validate the indexes somewhere?

Reply:
Yeah, that's the invariant enforced by the protocol (otherwise txs will not be included in the block at all):  https://github.com/stellar/stellar-protocol/blob/master/core/cap-0066.md#incorrect-entries-in-archivedentries-vector